_source/_posts/2025-11-04-user-entitlements-workshop.md

```diff
@@ -0,0 +1,1298 @@
+
+First, we'll create the interface for the `ResourceType` and define some strings. Open `okta-enterprise-ready-workshops/apps/api/src/scim-types.ts`. Add the interface for `IResourceType` above the `IListResponse` interface.
+
+
+```ts
+export interface IResourceType {
+  id?: string;
+  schemas: string[];
+  name: string; 
+  description?: string;
+  endpoint: string;
+  schema: string; 
+  meta: IMetadata;
+}
+```
+
+Notice the `IResourceType` doesn't extend from the `IScimResource` interface. For example, the SCIM standard doesn't require `id` for a resource type. Since the SCIM standard treats `ResourceType` as an exception case of `Resource`, we defined it separately without the relation instead of extending from `IScimResource`.
+
+When following the SCIM protocol, responses that list values, such as the list of roles or resource types, use the SCIM list response format.
+
+The `IListResource` interface must support `IOktaRole` and `IResourceType`. Using [generics](https://www.typescriptlang.org/docs/handbook/2/generics.html) and [union types](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types), we can support different list response objects . Update the `IListResource` to match the code below.
+
+```ts
+export interface IListResponse<T extends IScimResource | IResourceType> {
+  schemas: string[];
+  totalResults: number;
+  startIndex: number;
+  itemsPerPage: number;
+  Resources: T[];
+}
+```
+
+You'll see errors in the IDE and, if you're running the API, within the console output. No worries; we'll fix those errors soon!
+
+Resource types have a schema URN and use "ResourceType" as the `resourceType` string in the metadata. Add `SCHEMA_RESOURCE_TYPE` and edit `RESOURCE_TYPES` so your string constants section looks like the code below.
+
+```ts
+export const SCHEMA_OKTA_ROLE = 'urn:okta:scim:schemas:core:1.0:Role';
+export const SCHEMA_LIST_RESPONSE = 'urn:ietf:params:scim:api:messages:2.0:ListResponse';
+export const SCHEMA_RESOURCE_TYPE = 'urn:ietf:params:scim:schemas:core:2.0:ResourceType';
+export type RESOURCE_TYPES = 'Role' | 'ResourceType';
+```
+
+Open `okta-enterprise-ready-workshops/apps/api/src/entitlements.ts`. Let's fix the `IListResponse` error for the `/Roles` endpoint and specify the object type in the list, the `IOktaRole` type. The code building out the list changes to
+
+```ts
+  const listResponse: IListResponse<IOktaRole> = {
+    schemas: [SCHEMA_LIST_RESPONSE],
+    totalResults: roles.length,
+    startIndex,
+    itemsPerPage: recordLimit,
+    Resources: roles.map(role => ({
+      schemas: [SCHEMA_OKTA_ROLE],
+      id: role.id.toString(),
+      displayName: role.name
+    }))
+  };
+```
+
+You shouldn't see errors anymore! 🎉
+
+We have a new endpoint to add. Update the imports from the `./scim-types` file and declare a new route for resource types.
```

Comment:
```suggestion
We have a new endpoint to add. Update the imports from the `./scim-types` file and declare a new route for resource types in `entitlements.ts`:
```

at first I saw scim-types.ts and thought that was the file we were adding to. dumb mistake, easy fix.